### ISSUE TITLE
Neural HMM with a regression as emission process

### ISSUE THREAD

**Issue by prescientlyd** [jmschrei/pomegranate#898]
I noticed that the document of NeuralNetworkWrapper is based on classification as an emission process. I was wondering whether this method can also be used for a regression model as emission process. If yes, how should I change the code in the tutorial?

In tutorial, the code is:

distributions = [NeuralNetworkWrapper(nn, i, 28*28, 10) for i in range(10)]
model = HiddenMarkovModel.from_samples(distributions, 10, X_train_seqs, max_iterations=10, verbose=True)

But apparently, both i and number of classes are not applicable for a regression model.

**Comment by jmschrei** [comment]
The neural network has to be a classification model because it is predicting the probability of being in each hidden state in the HMM, which is a k-class classification problem for a HMM with k hidden states.

**Comment by prescientlyd** [comment]
Ah I see. I did not correctly understand the tutorial earlier. 

Now I have another question. The tutorial first trained a NN in which label of each image is used, the y_train_ohe in code below:
nn.fit(X_train, y_train_ohe, epochs=25, verbose=False, batch_size=256)

However, what should I do when I do not have this label information (y_train_ohe)? For example, for a POS task, I may not have the tag to start with.

The second question is that in this tutorial, the parameter of NN is fixed after NN is fit. What if I also want to further refine the parameters of NN together with the rest of the model (the HMM part)? Should I just do:
for layer in nn.layers:
    layer.trainable = True

Anything else I need to change to update the parameters of NN? 

Thanks!

**Comment by jmschrei** [comment]
Theoretically, it's not necessary to pre-train the neural network; however, sticking gradient descent inside expectation-maximization can be very difficult to train. The gist is that the neural network is trained in the same way that the distributions in the HMM would be: at each iteration of EM, you use the expectation step to generate pseudo-labels for each example and then in the maximization step you use those labels to fit the distribution. When it's a neural network, you generate pseudo-labels as before, and use those as targets in one step of gradient descent. In practice, the learning rate should be set to be pretty low so that you don't overshoot the target. This fits the transition matrix and the neural network jointly. In the tutorial I freeze the neural network because it's waaaay easier to just fit the transition matrix when the emissions are fixed but there's no theoretical reason that's necessary.

**Comment by prescientlyd** [comment]
This makes perfect sense. However, for the case in which the latent states are not observable and that we cannot pre-train the neural network, what is the best strategy I should do to initialize the parameters of the neural network?

**Comment by jmschrei** [comment]
Theoretically, you're not required to have a good initialization. However, you might try pre-training the network as an autoencoder.

**Comment by prescientlyd** [comment]
This is a great idea! So in this case, if I setup 10 latent states for the HMM, in autoencoder, I should also set the size of representation to 10 right?

**Comment by jmschrei** [comment]
There are a variety of ways you could do it. That's one way, though the internal dimensions of an autoencoder might not correspond to distinct classes. You could also train an autoencoder, take and freeze the trained encoder, add an unfrozen 10 node layer on top and just train that with the transition matrix.
